eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+            HashMap<String, TrackedSourcedEventMessage> candidateMessagesToReturn = new HashMap<>();
+
+            peekForMessages(candidateMessagesToReturn);
+
+            //select message to return from candidates
+            final Optional<Map.Entry<String, TrackedSourcedEventMessage>> chosenMessage =
+                    candidateMessagesToReturn.entrySet().stream().min((m1, m2) -> comparator
+                            .compare(m1.getValue(), m2.getValue()));
+
+            // Ensure the chosen message is actually consumed from the stream
+            return chosenMessage.map(e -> {
+                String streamId = e.getKey();
+                TrackedEventMessage<?> message = e.getValue();
+                try {
+                    return new GenericTrackedEventMessage<>(trackingToken.advancedTo(streamId, message.trackingToken()),
+                                                            messageSources.get(streamId).nextAvailable());
+                } catch (InterruptedException ex) {
+                    logger.warn("Thread Interrupted whilst consuming next message", ex);
+                }
+                return null;
+            }).orElse(null);
+        }
+
+        /**
+         * @param timeout the maximum number of time units to wait for messages to become available. This time
+         *                is spent polling on one particular stream (specified by {@link Builder#configureLongPollingSource(String)}
+         *                or by default to last stream provide to {@link Builder#addMessageSource(String,
+         *                StreamableMessageSource)}.
+         *                This stream long polls for a fraction of the specified duration before looping through the
+         *                sources
+         *                again repeating until the duration has been met. This ensure the highest chance of a
+         *                consumable
+         *                message being found.
+         * @param unit    the time unit for the timeout.
+         * @return true if any stream has an available message. Otherwise false.
+         *
+         * @throws InterruptedException when the thread is interrupted before the indicated time is up.
+         */
+        @Override
+        public boolean hasNextAvailable(int timeout, TimeUnit unit) throws InterruptedException {
+            long deadline = System.currentTimeMillis() + unit.toMillis(timeout);
+            int longPollTime = timeout / 10;
+
+            while (System.currentTimeMillis() < deadline) {
+                Iterator<Map.Entry<String, BlockingStream<TrackedEventMessage<?>>>> it = messageSources.entrySet()
+                                                                                                       .iterator();
+
+                while (it.hasNext()) {
+                    Map.Entry<String, BlockingStream<TrackedEventMessage<?>>> current = it.next();
+
+                    if (it.hasNext()) {
```

Comment:
For clarity, I think it's good to point out that this if-else block is meant to diverge between all the message sources and the last message sources as the last should be treated differently.

Either through a comment or a named method I'd say; I am fine with both.